relay/tests/test_dynamic_sampling.rs

```diff
@@ -0,0 +1,326 @@
+use serde_json::json;
+use uuid::Uuid;
+
+use relay_base_schema::project::ProjectId;
+use relay_sampling::config::RuleType;
+use relay_test::{
+    create_error_item, mini_sentry::MiniSentry, new_sampling_rule, relay::Relay, Envelope, Outcome,
+    ProjectState,
+};
+
+/// Tests that when sampling is set to 0% for the trace context project the events are removed.
+#[test]
+fn test_it_removes_events() {
+    let project_state = ProjectState::new()
+        // add a sampling rule to project config that removes all transactions (sample_rate=0)
+        .add_basic_sampling_rule(RuleType::Transaction, 0.0)
+        .set_transaction_metrics_version(1);
+
+    let public_key = project_state.public_key();
+    let sentry = MiniSentry::new().add_project_state(project_state);
+    let relay = Relay::builder(&sentry).enable_outcomes().build();
+
+    // create an envelope with a trace context that is initiated by this project (for simplicity)
+    let envelope = Envelope::new()
+        .add_basic_transaction(None)
+        .set_basic_trace_info(public_key);
+
+    // send the event, the transaction should be removed.
+    relay.send_envelope(envelope);
+
+    // the event should be removed by Relay sampling
+    sentry.captured_envelopes().wait(2).assert_empty();
+
+    sentry
+        .captured_outcomes()
+        .wait_for_outcome(30)
+        .assert_outcome_qty(1)
+        .assert_all_outcome_id(Outcome::FILTERED)
+        .assert_all_outcome_reasons("Sampled:1");
```

Comment:
Could this be more generic? The functions `assert*outcome*` seem too specific, would something like this make sense instead:

```suggestion
    let outcomes = 
        .captured_items::<Outcome>()
        .wait_for(30);
    let outcome = outcomes.next();
    assert_eq!(outcome.id, Outcome::Filtered);
    assert_eq!(outcome.reason, "Sampled:1");
    assert!(outcomes.next().is_none());
```

Reply:
well yes, could be more generic as the current python way of doing it, I liked this way of chaining asserts together though so it depends on the preference. I think we can discuss it as a team